cmake/OpenCVDetectCUDA.cmake

```diff
@@ -101,28 +130,20 @@ if(CUDA_FOUND)
       set(__cuda_arch_bin "3.2")
       set(__cuda_arch_ptx "")
     elseif(AARCH64)
-      execute_process( COMMAND "${CUDA_NVCC_EXECUTABLE}" ${CUDA_NVCC_FLAGS} "${OpenCV_SOURCE_DIR}/cmake/checks/OpenCVDetectCudaArch.cu" "--run"
-                       WORKING_DIRECTORY "${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/CMakeTmp/"
-                       RESULT_VARIABLE _nvcc_res OUTPUT_VARIABLE _nvcc_out
-                       ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
+      ocv_detect_native_cuda_arch(_nvcc_res _nvcc_out)
       if(NOT _nvcc_res EQUAL 0)
         message(STATUS "Automatic detection of CUDA generation failed. Going to build for all known architectures.")
         set(__cuda_arch_bin "5.3 6.2 7.2")
       else()
         set(__cuda_arch_bin "${_nvcc_out}")
-        string(REPLACE "2.1" "2.1(2.0)" __cuda_arch_bin "${__cuda_arch_bin}")
       endif()
       set(__cuda_arch_ptx "")
     else()
-      if(CUDA_VERSION VERSION_LESS "9.0")
-        set(__cuda_arch_bin "2.0 3.0 3.5 3.7 5.0 5.2 6.0 6.1")
-      elseif(CUDA_VERSION VERSION_LESS "10.0")
-        set(__cuda_arch_bin "3.0 3.5 3.7 5.0 5.2 6.0 6.1 7.0")
-      else()
-        set(__cuda_arch_bin "3.0 3.5 3.7 5.0 5.2 6.0 6.1 7.0 7.5")
-      endif()
+      set(__entire_cuda_arch_bin "${_arch_fermi} ${_arch_kepler} ${_arch_maxwell} ${_arch_pascal} ${_arch_volta} ${_arch_turing}")
+      ocv_filter_available_architecture(__cuda_arch_bin ${__entire_cuda_arch_bin})
```

Comment:
It would be nice to split line above (so future modification and history would be straightforward) and use `${ARGN}` in `ocv_filter_available_architecture` macro
```
ocv_filter_available_architecture(__cuda_arch_bin
    ${_arch_fermi}
    ${_arch_kepler}
    ${_arch_maxwell}
    ${_arch_pascal}
    ${_arch_volta}
    ${_arch_turing}
)
```

```.diff
-macro(ocv_filter_available_architecture result_list TARGET_ARCH_LIST)
+macro(ocv_filter_available_architecture result_list)
-  string(REPLACE " " ";" CC_LIST ${TARGET_ARCH_LIST})
+  set(CC_LIST ${ARGN})
```